include/seqan3/search/bloom_filter/bloom_filter.hpp

```diff
@@ -0,0 +1,386 @@
+     * \param size The bit vector size.
+     * \param funs The number of hash functions. Default 2. At least 1, at most 5.
+     *
+     * \attention This constructor can only be used to construct **uncompressed** Bloom Filters.
+     *
+     * \details
+     *
+     * ### Example
+     *
+     * \include test/snippet/search/bloom_filter/bloom_filter_constructor.cpp
+     */
+    bloom_filter(seqan3::bin_size size,
+                 seqan3::hash_function_count funs = seqan3::hash_function_count{2u})
+    //!\cond
+        requires (data_layout_mode == data_layout::uncompressed)
+    //!\endcond
+    {
+        size_ = size.get();
+        hash_funs = funs.get();
+
+        if (hash_funs == 0 || hash_funs > 5)
+            throw std::logic_error{"The number of hash functions must be > 0 and <= 5."};
+        if (size_ == 0)
+            throw std::logic_error{"The size of a bloom filter must be > 0."};
+
+        hash_shift = std::countl_zero(size_);
+        data = sdsl::bit_vector(size_);
+    }
+
+    /*!\brief Construct a compressed Bloom Filter.
+     * \param[in] bf The uncompressed seqan3::bloom_filter.
+     *
+     * \attention This constructor can only be used to construct **compressed** Bloom Filters.
+     *
+     * \details
+     *
+     * ### Example
+     *
+     * \include test/snippet/search/bloom_filter/bloom_filter_constructor_compressed.cpp
+     */
+    bloom_filter(bloom_filter<data_layout::uncompressed> const & bf)
+    //!\cond
+        requires (data_layout_mode == data_layout::compressed)
+    //!\endcond
+    {
+        std::tie(size_, hash_shift, hash_funs) =
+            std::tie(bf.size_, bf.hash_shift, bf.hash_funs);
+
+        data = sdsl::sd_vector<>{bf.data};
+    }
+    //!\}
+
+    /*!\name Modifiers
+     * \{
+     */
+    /*!\brief Inserts a value into the bloom filter.
+     * \param[in] value The raw numeric value to process.
+     *
+     * \attention This function is only available for **uncompressed** Bloom Filters.
+     *
+     * \details
+     *
+     * ### Example
+     *
+     * \include test/snippet/search/bloom_filter/bloom_filter_emplace.cpp
+     */
+    void emplace(size_t const value) noexcept
+    //!\cond
+        requires (data_layout_mode == data_layout::uncompressed)
+    //!\endcond
+    {
+        for (size_t i = 0; i < hash_funs; ++i)
+        {
+            size_t idx = hash_and_fit(value, hash_seeds[i]);
+            assert(idx < data.size());
+            data[idx] = 1;
+        };
+    }
+
+    /*!\brief Clears the complete Bloom Filter data.
+     *
+     * \attention This function is only available for **uncompressed** Bloom Filters.
+     *
+     * \details
+     *
+     * ### Example
+     *
+     * \include test/snippet/search/bloom_filter/bloom_filter_clear.cpp
+     */
+    void clear() noexcept
+    //!\cond
+        requires (data_layout_mode == data_layout::uncompressed)
+    //!\endcond
+    {
+        data = sdsl::bit_vector(size_);
```

Comment:
```suggestion
        data.clear();
```

The move assignment probably involves allocating new memory.
Clear should just be linear in the size.

Reply:
I don't think this is appropriate, as - according to the behavior of std containers, `data.clear()` would also reset the size of the bitvector to 0. I think for the `bloom_filter::clear()`, we want to set all entries to 0 rather than changing the vector size, don't we!?  
  
Thus, maybe the best method would be using `std::fill`, which should guarantee to prevent reallcation and has also linear time:
`std::fill(data.begin(), data.end(), 0);`  
  
In this case, I don't think we need to benchmark, this should be the better option to reinitialization.

Reply:
> I don't think this is appropriate, as - according to the behavior of std containers - `data.clear()` would also reset the size of the bitvector to 0. I think for the `bloom_filter::clear()`, we want to set all entries to 0 rather than changing the vector size, don't we!?
> 
> Thus, maybe the best method would be using `std::fill`, which should guarantee to prevent reallcation and has also linear time:
> `std::fill(data.begin(), data.end(), 0);`
> 
> In this case, I don't think we need to benchmark, this should be the better option to reinitialization.

Yes, you are right, please use `std::fill`. This is an error I like to make :)

Reply:
When using `std::fill`, a benchmark would still be interesting, because the iterators of the `sdsl::bit_vector` are really slow.

Reply:
> 
> 
> When using `std::fill`, a benchmark would still be interesting, because the iterators of the `sdsl::bit_vector` are really slow (I'd figure they are not slower than allocating memory, but who knows...)
> 
> Doesn't have to be very extensive, just running the benchmark you proposed once with each implementation.

You are absolutely right, the `std::fill` version is MUCH slower, even with optimizations turned on (where I'd expect it to be translated to memset). Though, I don't know if a reallocation took place in the reinit version.  
I will try a manual memset version and see whether this improves something.

Reply:
Can you try 
```cpp
    sdsl::util::_set_zero_bits(data);
```

this uses the data pointer, and should be reasonably fast

Reply:
I ran some different options, after every change also running the tests to ensure that it does what it should (at least within the tested scope):
```
// manual memset; most efficient, but huge trouble if something is going wrong
memset(data.data(), 0, ((data.bit_size() + 63) >> 3));
---
/32768/2/1000            36.5 ns         36.4 ns     19418859
/1048576/2/1000          2245 ns         2240 ns       310369


// sdsl::util::_set_zero_bits; some loss of efficiency in my setup, but still doing quite well
sdsl::util::_set_zero_bits(data);
---
/32768/2/1000             201 ns          200 ns      3419712
/1048576/2/1000          6202 ns         6036 ns       111505


// std::fill; very inefficient
std::fill(data.begin(), data.end(), 0);
---
/32768/2/1000           39396 ns        39305 ns        18462
/1048576/2/1000       1259445 ns      1235777 ns          560


// Reinitialization; nearly as efficient as using memset
data = sdsl::bit_vector(size_in_bits, 0);
---
/32768/2/1000            62.7 ns         62.6 ns     11083526
/1048576/2/1000          2340 ns         2333 ns       300935


// Creating new bit vector to enforce reallocation; 
// though, the bitvector is pretty small in this example, so there is no huge impact
sdsl::bit_vector data2(size_in_bits, 0);
data = std::move(data2);
---
/32768/2/1000            63.4 ns         63.3 ns     10542822
/1048576/2/1000          2376 ns         2285 ns       309440
```

Thus, I see `memset` or `_set_zero_bits` as the best options, the first being more efficient, the second being more safe to use. If reallocation is not too much of a problem, reinitialization (as my original implementation) could also still be an option. I leave that decision to you.

Reply:
It would be ok for me to use `memset`.
We can also use `_set_zero_bits` and change it in the SDSL (which I'll do anyway).

If you need `clear` in your application, you may want the faster version, otherwise we can use the SDSL version.

Reply:
I think, I would go for the `_set_zero_bits` version and wait for the SDSL changes.  
I don't feel THAT comfortable with direct memory access, especially when I am not 100% sure about the details of the underlying data structure (which is the case for the SDSL bit vectors).  
Additionally, I don't expect anyone to call `clear()` too often, it should usually have only minor impact on a global perspective. So I think it's fine to use the SDSL for now and wait for the change there such that someone who is more familiar with the details of the data structure can have a look (I guess, there is a reason why they didn't directly implement the memset there, but only put a comment).

Reply:
I did this benchmark https://gist.github.com/eseiler/21023865b43b6567a39fe250cf43c027
```
---------------------------------------------------------------------
Benchmark                           Time             CPU   Iterations
---------------------------------------------------------------------
set_to_zero_via_fill<1>        852937 ns       852939 ns          850
set_to_zero_via_fill<8>         11022 ns        11022 ns        62989
set_to_zero_via_fill<16>        23139 ns        23139 ns        30210
set_to_zero_via_fill<32>        49874 ns        49874 ns        14071
set_to_zero_via_fill<64>       127919 ns       127916 ns         5333

set_to_zero_via_memset<1>         999 ns          999 ns       686000
set_to_zero_via_memset<8>       11211 ns        11211 ns        61638
set_to_zero_via_memset<16>      23167 ns        23168 ns        30146
set_to_zero_via_memset<32>      49401 ns        49401 ns        14111
set_to_zero_via_memset<64>     125041 ns       125041 ns         5711

set_to_zero_via_util<1>          5798 ns         5798 ns       127256
set_to_zero_via_util<8>         46142 ns        46143 ns        15099
set_to_zero_via_util<16>        93289 ns        93289 ns         7986
set_to_zero_via_util<32>       190294 ns       190295 ns         3674
set_to_zero_via_util<64>       380571 ns       380572 ns         1692

set_to_zero_via_assign<1>        5818 ns         5818 ns       111745
set_to_zero_via_assign<8>       46895 ns        46895 ns        14814
set_to_zero_via_assign<16>      91859 ns        91859 ns         7355
set_to_zero_via_assign<32>     203704 ns       203704 ns         3648
set_to_zero_via_assign<64>     722713 ns       722716 ns          933
```

Reply:
Interesting, so this looks like `std::fill` actually optimizes to `memset`, but only for word size>=8, or even only for multiples of 8 (would be interesting to see something like `set_to_zero_via_fill<30>`, but I guess it's just not possible to optimize this for non-multiples of 8 which actually would make sense). And assignment looks actually worse than util, but mainly for larger words (which would be irrelevant for bitvectors).  
So I guess, the main conclusion would be the same as derived from my benchmarks (using util as secure, or memset as most efficient solution), or is there anything I oversee!?

Reply:
We're doing https://github.com/xxsds/sdsl-lite/pull/86 now,

I guess fill can optimize for multiples of `8`, for `1` it needs to iterate (=slow).
It's also slow for, e.g., `30`.

The other versions don't care about the width, because they work directly on the data.